src/pce/pce.jl

```diff
@@ -0,0 +1,553 @@
+                continue
+            end
+            indices_i += 1
+            if indices_i > indices_size
+                return res
+            end
+        end
+    end
+    res
+end
+
+"""
+$(TYPEDEF)
+`$(FUNCTIONNAME)` represents multivariate orthogonal polynomial bases formed as the tensor
+product of univariate `PolyChaos.AbstractOrthoPoly`s.
+
+`$(FUNCTIONNAME)` is similar to `PolyChaos.MultiOrthoPoly`, but allows different degree
+truncation for univariate bases. Multi-indices of `$(FUNCTIONNAME)` are stored in the
+columns as Julia is column-major, while multi-indices are rows in the index matrix of
+`PolyChaos.MultiOrthoPoly`.
+
+By default, the sum of multi-indices are restricted to the maximum degree among the
+univariate bases.
+
+# Fields
+$(TYPEDFIELDS)
+"""
+struct TensorProductOrthoPoly{OP <: Union{AbstractOrthoPoly, AbstractCanonicalOrthoPoly}} <:
+       AbstractOrthoPoly{ProductMeasure, AbstractQuad{Float64}}
+    "The degree truncation of each univariate orthogonal polynomials."
+    deg::Vector{Int}
+    "Multi-indices in the columns."
+    ind::Matrix{Int}
+    "Product measure."
+    measure::ProductMeasure
+    "Univariate orthogonal polynomials."
+    uni::Vector{OP}
+end
+"""
+$(TYPEDSIGNATURES)
+"""
+function TensorProductOrthoPoly(ops::AbstractVector{
+                                                    <:Union{AbstractOrthoPoly,
+                                                            AbstractCanonicalOrthoPoly}})
+    if any(op.quad isa EmptyQuad for op in ops)
+        throw(InconsistencyError("at least one quadrature rule missing"))
+    end
+    degrees = deg.(ops)
+    ind = grlex(degrees)
+    measures = [op.measure for op in ops]
+    w(t) = prod(m.w(t) for m in measures)
+    measure = ProductMeasure(w, measures)
+    TensorProductOrthoPoly(degrees, ind, measure, Vector(ops))
+end
+
+PolyChaos.dim(tpop::TensorProductOrthoPoly) = size(tpop.ind, 2)
+
+function PolyChaos.computeTensorizedSP(dim::Integer, tpop::TensorProductOrthoPoly)
+    computeTensorizedSP(dim, tpop.uni, transpose(tpop.ind))
+end
+
+"""
+$(TYPEDSIGNATURES)
+Construct a `$(FUNCTIONNAME)` which is used to compute and store the results of inner
+products
+```math
+⟨ϕ_{i_1}ϕ_{i_2}⋯ϕ_{i_{m-1}},ϕ_{i_m}⟩
+```
+where ``m`` is input argument `dim`.
+"""
+function PolyChaos.Tensor(dim::Int, tpop::TensorProductOrthoPoly)
+    tensor_entries = computeTensorizedSP(dim, tpop)
+    getfun(ind) = getentry(ind, tensor_entries, transpose(tpop.ind), dim)
+    Tensor(dim, tensor_entries, getfun, tpop)
+end
+
+"""
+$(TYPEDEF)
+Represents product of orthogonal polynomial basis functions
+``Ψ_{\\vec α_1} Ψ_{\\vec α_2} ⋯ Ψ_{\\vec α_m}``.
+
+This is used to combine product of basis functions into one single object and conveniently
+take inner product tensor ``⟨⋅, Ψ_{\\vec β}⟩``.
+
+# Fields
+$(TYPEDFIELDS)
+"""
+struct BasisProduct
+    "The indices of the multi-indices ``\\vec α_1, \\vec α_2, …, \\vec α_m``."
+    idx::Vector{Int}
+end
```

Comment:
With `Vector`, `BasisProduct([3, 5, 4, 3, 4, 4])` and `BasisProduct([4, 3, 4, 4, 5, 3])` are not the same, but they are both mathematically equivalently to $Ψ_3^2Ψ_4^3Ψ_5$. So this could be changed to be represented as `Dict(3 => 2, 4 => 3, 5 => 1)` just like [`SymbolicUtils.Mul`](https://github.com/JuliaSymbolics/SymbolicUtils.jl/blob/490011e2dbc62b53ffbdb9e9b3f27dafa9b2f74d/src/types.jl#L39). 

However, as of Julia v1.8.2, the [constructor method](https://github.com/JuliaLang/julia/blob/36034abf26062acad4af9dcec7c4fc53b260dbb4/base/dict.jl#L88-L91) of `Dict()` allocates 3 `Vector`s of size `n = 16` by default. We probably rarely have product of that many distinct terms in practice. The size of multi-indices $\vec α$ has exponential space complexity, so the constant `3 * n = 3 * 16 = 48` is quite significant. 

Furthermore, [`Dict`](https://github.com/JuliaLang/julia/blob/36034abf26062acad4af9dcec7c4fc53b260dbb4/base/dict.jl#L78) is a mutable struct, but we never need to mutate it for our purpose except possibly calling [`mergewith(+, p, q)`](https://docs.julialang.org/en/v1.8/base/collections/#Base.mergewith) for combining 2 objects in `*`.

[`ImmutableDict`](https://docs.julialang.org/en/v1.8/base/collections/#Base.ImmutableDict) is also one option.

I leave it as simple `Vector` for now and will modify the data structure later if necessary.

Reply:
I agree with that solution. I am not sure the downside of having a non-unique representation is all that critical. We can always pick the ordered list into a canonical representation if we need it, say for comparison. That seems rarely important, at least for the purpose of PCE.

Reply:
Right. OK.

Reply:
Resolved by doing "Merge" of MergeSort in `*`.